src/bbtautau/postprocessing/postprocessing.py

```diff
@@ -971,6 +1014,65 @@ def bbtautau_assignment(
         sample.tt_mask = bbtt_masks["tt"]
 
 
+def _get_lepton_mask(sample, lepton_type: str, dR_cut: float) -> np.ndarray:
+    """
+    Calculates a boolean mask to select the highest-pT lepton of a given type
+    that is within a dR_cut of the ttFatJet.
+
+    Returns a boolean array with at most one 'True' per row (event).
+    """
+    # Use .to_numpy() for efficient computation.
+    # The jet's (N,) arrays are reshaped to (N, 1) to broadcast correctly
+    # against the lepton's (N, M) arrays, where M is the number of leptons.
+
+    lepton_eta = sample.get_var(f"{lepton_type}Eta")
+    lepton_phi = sample.get_var(f"{lepton_type}Phi")
+    jet_eta = sample.get_var("ttFatJetEta")[:, np.newaxis]
+    jet_phi = sample.get_var("ttFatJetPhi")[:, np.newaxis]
+
+    # print(sample.sample.label)
+    # print(np.sum(lepton_eta != PAD_VAL, axis=0)/len(lepton_eta))
+
+    # 1. Calculate dR for all leptons and create an initial mask
+    dR = np.sqrt((lepton_eta - jet_eta) ** 2 + (lepton_phi - jet_phi) ** 2)
+    initial_mask = dR < dR_cut
+
+    # 2. Find events that have at least one passing lepton
+    events_with_any_pass = initial_mask.any(axis=1)
+
+    # 3. Find the index of the *first* (highest-pT) passing lepton in each event
+    # For events where none pass, np.argmax incorrectly returns 0.
+    first_pass_idx = np.argmax(initial_mask, axis=1)
+
+    # 4. Create the final mask, initialized to all False
+    final_mask = np.zeros_like(initial_mask, dtype=bool)
+
+    # 5. Use advanced indexing to set a single 'True' in the correct spot.
+    # This clever line only sets final_mask[i, j] to True if events_with_any_pass[i]
+    # is True, effectively ignoring the incorrect argmax result for non-passing events.
+    row_indices = np.arange(len(final_mask))
+    final_mask[row_indices, first_pass_idx] = events_with_any_pass
+
+    return final_mask
+
+
+def leptons_assignment(
```

Comment:
[For future] would be good to have control plots for how many leptons pass this "assignment", for signals and backgrounds.